### ISSUE TITLE
Options for padding

### ISSUE THREAD

**Issue by flyIchtus** [kymatio/kymatio#859]
Working with scattering for weather forecasts image analysis, I was wondering what influence the type of padding did have on results.

The reflection padding implemented by default seems to produce high gradients on the boundaries. After a few subsamplings, the large values on the boundary gain a certain weight in the spatial average, that could lead to estimation biases.

Especially, I want to estimate ratios of

Question 1 : have the topic already been discussed ?

Question 2 : what about providing numpy padding modes as an option to scattering ?

After a fork and 1 or 2 adds on numpy frontend/backend, one comes to the following  code :

**Comment by MuawizChaudhary** [comment]
[This discussion ](https://github.com/kymatio/kymatio/discussions/752) and PRs #727, #734, #738 might be references to discussions of a requested feature of different types of padding. I don't know if we have plans for adding different types of padding. 

@lostanlen is in the process of refactoring the padding module/operation from the 1D scattering.

@edouardoyallon do you think its a good idea to add different types of padding in 2d, like reflective or constant or what not? It should be easy to do I think, but correctness/comparison to a base implementation might be an issue

**Comment by lostanlen** [comment]
Hello @flyIchtus ,

Right now, what you're doing is indeed the only way to have custom padding/unpadding in Kymatio 2D: pass `pre_pad=False` and roll up your own implementation, `order2_ind` in your case.

But i agree that this is not satisfactory and ideally you should be able to pass a single keyword argument at runtime. Not:
```python
scattering=Scattering2D(4,data.shape,8, pad_type=pad_type)
results_order2=scattering(data)
...
```

But

```python
scattering=Scattering2D(4,data.shape,8)
results_order2=scattering(data, pad_type=pad_type)
...
```
(the keyword `pad_type` is up for debate. It's not a "type" in the CS sense of the word)

In other words, the `Scattering2D` object shouldn't need to know about what type of padding you'll want to apply later. From a user's perspective, this will allow you to have a single object computing scattering transforms with several kinds of padding. When that plan is ready, all your code will rewrite as

```python
scattering = Scattering2D(J=4, shape=data.shape, L=8)
S2_symm = scattering(data, pad_type='symmetric')
S2_reflect = scattering(data, pad_type='reflect')
```

Thoughts?

**Comment by flyIchtus** [comment]
OK, sorry for the long delay. Thank you both for pointing me to useful discussions.

Great proposal @lostanlen , having flexible padding options for the same scattering object really more elegant (and lean).
Looking at #727, the main issue for implementation seems to be the diversity of backends.

Agree that pad_mode seems better than pad_type.

As for validation and correctness issue @MuawizChaudhary  : 

A starting point could be something like https://arxiv.org/abs/2010.02178
and showing what bias the different padding modes introduce, especially on intermediate activations (my modest plot upwards was a push in this direction). 

Hoping for "perfect" 2D padding, that would allow to forget boundary effects, seems unreasonable however. 

Anyway, on texture datasets, it would be interesting to (statistically) compare

-  the effect of (M,N)-padding on (H,W) images
-  the effect of no padding on (H+M, W+N) images

I might be able to have a try on my weather dataset (which, strictly speaking, is not a texture but incorporate some). I keep you updated on this issue 1) if you are interested and 2) if i can get it done in the coming weeks.

**Comment by lostanlen** [comment]
#926 has added zero-padding to 1D backends

**Comment by lostanlen** [comment]
#944
